server/s3_post_complete_action.cc

```diff
@@ -505,76 +596,56 @@ void S3PostCompleteAction::add_object_oid_to_probable_dead_oid_list() {
-    // TODO add probable part list index delete record.
-
-    // Mark old object if any for probable deletion.
-    // store old object oid
-    if (old_object_oid.u_hi || old_object_oid.u_lo) {
-      assert(!old_oid_str.empty());
-      assert(!new_oid_str.empty());
-
-      // prepending a char depending on the size of the object (size based
-      // bucketing
-      // of object)
-
-      S3CommonUtilities::size_based_bucketing_of_objects(old_oid_str,
-                                                         object_size);
-
-      s3_log(S3_LOG_DEBUG, request_id,
-             "Adding old_probable_del_rec with key [%s]\n",
-             old_oid_str.c_str());
-
-      // key = oldoid + "-" + newoid
-      std::string old_oid_rec_key = old_oid_str + '-' + new_oid_str;
-
-      old_probable_del_rec.reset(new S3ProbableDeleteRecord(
-          old_oid_rec_key, {0ULL, 0ULL}, multipart_metadata->get_object_name(),
-          old_object_oid, old_layout_id,
-          bucket_metadata->get_object_list_index_oid(),
-          bucket_metadata->get_objects_version_list_index_oid(),
-          multipart_metadata->get_version_key_in_index(),
-          false /* force_delete */
-          ));
-      // backgrounddelete will delete this entry if current object metadata has
-      // moved on
+    add_part_object_to_probable_dead_oid_list(new_object_metadata,
+                                              new_parts_probable_del_rec_list);
+    std::map<std::string, std::string> kv_list;
+    for (auto& probable_rec : new_parts_probable_del_rec_list) {
+      kv_list[probable_rec->get_key()] = probable_rec->to_json();
+      new_obj_oids.push_back(probable_rec->get_current_object_oid());
+    }
+    if (!kv_list.empty()) {
+      // S3 Background delete will delete new object parts, when multipart
+      // metadata has been deleted
       motr_kv_writer->put_keyval(
-          global_probable_dead_object_list_index_oid, old_oid_rec_key,
-          old_probable_del_rec->to_json(),
+          global_probable_dead_object_list_index_oid, kv_list,
           std::bind(&S3PostCompleteAction::
                          add_object_oid_to_probable_dead_oid_list_success,
                     this),
           std::bind(&S3PostCompleteAction::
                          add_object_oid_to_probable_dead_oid_list_failed,
                     this));
+    } else {
+      next();
+    }
+  } else {
+    // TODO add probable part list index delete record.
+    // Mark old object if any for probable deletion.
+    // store old object oid
+    if (old_object_oid.u_hi || old_object_oid.u_lo) {
+      assert(!old_oid_str.empty());
+      assert(!new_oid_str.empty());
+      add_part_object_to_probable_dead_oid_list(
+          object_metadata, old_parts_probable_del_rec_list, true);
+      std::map<std::string, std::string> kv_list;
+      for (auto& probable_rec : old_parts_probable_del_rec_list) {
+        kv_list[probable_rec->get_key()] = probable_rec->to_json();
```

Comment:
Can we be sure, that all keys would be unique?

Reply:
yes. key here would be unique...because they are OIDs